### ISSUE TITLE
rms_norm autocast fp32 policy (2.12) materializes standalone cast kernels under torch.compile when dtypes already match. Peak-memory and throughput regression for bf16 models

### ISSUE THREAD

**Issue by abhishek-patnia-nu** [pytorch/pytorch#188955]
### 🐛 Describe the bug

torch 2.12 registered `rms_norm` under autocast's `fp32` cast policy ([#174824](https://github.com/pytorch/pytorch/pull/174824)). The policy applies unconditionally, including when input and weight are already both bf16. On torch 2.11 the op had no autocast registration, so in this configuration it ran bf16-in/bf16-out and autocast played no role.

```python
# torch 2.12.1, bf16 input AND bf16 weight:
with torch.autocast('cuda', dtype=torch.bfloat16):
    y = norm(x_bf16)     # y.dtype == torch.float32  <- promoted regardless
with torch.autocast('cuda', enabled=False):
    y = norm(x_bf16)     # y.dtype == torch.bfloat16
```

The promotion inserts an fp32 upcast of the activation before every norm and leaves an fp32 output
that is downcast by the next `lower_precision_fp` op. Under `torch.compile`, these casts are graph
nodes with multiple consumers (the promoted input feeds both the statistic reduction and the
elementwise normalize). Some fuse into the generated norm kernels, but a large fraction are emitted
as standalone `triton_poi_fused__to_copy*` kernels, each writing a full-size buffer — a full fp32
copy of the activation per cast.

The promotion also provides no numerical benefit in this configuration: `rms_norm` upcasts
internally regardless (`rms_norm_composite`: `input.to(opmath_t)`; generated kernels accumulate the
statistic in fp32 registers), and upcasting a bf16-stored weight recovers no information.

At scale this is a significant delta. 

### Minimal reproduction

A single `Linear(4096, 4096)` followed by `RMSNorm(4096)`, model cast
to bf16, compiled, forward+backward under bf16 autocast, input (8, 4096, 4096). The only
difference between the two modes is whether the norm call is wrapped in
`torch.autocast(enabled=False)`:

<details>
<summary>repro.py</summary>

```python
# ruff: noqa
"""Repro: rms_norm autocast fp32 policy materializes cast kernels under
torch.compile when input and weight are already bf16 (torch 2.12 vs 2.11).

Minimal setup: Linear -> RMSNorm, model cast to bf16, compiled,
forward+backward under bf16 autocast.

torch 2.11: rms_norm has no autocast registration; with matched bf16 dtypes
the compiled step emits no standalone copy kernels.
torch 2.12: rms_norm is registered under autocast's fp32 cast policy
(pytorch#174824), which promotes unconditionally; the promotion casts
materialize as standalone `triton_poi_fused__to_copy*` kernels and raise
peak memory.

Run (single GPU):
    python repro.py            # autocast-wrapped norm (regression path)
    python repro.py exempt     # norm under autocast(enabled=False) (workaround)
"""

import sys
import time

import torch
import torch.nn as nn

D, BATCH, SEQ, ITERS = 4096, 8, 4096, 20
EXEMPT = len(sys.argv) > 1 and sys.argv[1] == "exempt"


class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(D, D, bias=False)
        self.norm = nn.RMSNorm(D, eps=1e-6)

    def forward(self, x):
        h = self.linear(x)
        if EXEMPT:
            with torch.autocast(x.device.type, enabled=False):
                return self.norm(h)
        return self.norm(h)


model = Net().to("cuda", torch.bfloat16)
model = torch.compile(model)


def step():
    x = torch.randn(BATCH, SEQ, D, device="cuda", dtype=torch.bfloat16, requires_grad=True)
    with torch.autocast("cuda", dtype=torch.bfloat16):
        out = model(x)
    out.float().sum().backward()


for _ in range(3):  # warmup + compile
    step()
torch.cuda.synchronize()
torch.cuda.reset_peak_memory_stats()

t0 = time.perf_counter()
for _ in range(ITERS):
    step()
torch.cuda.synchronize()
step_ms = (time.perf_counter() - t0) / ITERS * 1000
peak_gib = torch.cuda.max_memory_allocated() / 2**30

from torch.profiler import ProfilerActivity, profile

with profile(activities=[ProfilerActivity.CUDA]) as prof:
    step()
    torch.cuda.synchronize()

kernels = copies = 0
for ev in prof.key_averages():
    if getattr(ev, "device_time_total", 0) <= 0 or ev.key.startswith(("aten::", "Prof", "cuda")):
        continue
    kernels += ev.count
    if "to_copy" in ev.key or "convert_element_type" in ev.key:
        copies += ev.count

print(
    f"torch={torch.__version__} mode={'exempt' if EXEMPT else 'autocast'} "
    f"peak_gib={peak_gib:.2f} step_ms={step_ms:.2f} kernels/step={kernels} cast_kernels/step={copies}"
)
```

</details>

```bash
python repro.py            # autocast-wrapped norm (regression path)
python repro.py exempt     # norm under autocast(enabled=False) (workaround path)
```

Output on torch 2.12.1+cu130, single B200 (the LayerNorm line is the same script with
`nn.LayerNorm` substituted):

```
torch=2.12.1+cu130 RMSNorm   mode=autocast peak_gib=1.97 step_ms=7.01 kernels/step=15 cast_kernels/step=3
torch=2.12.1+cu130 RMSNorm   mode=exempt   peak_gib=1.47 step_ms=7.43 kernels/step=16 cast_kernels/step=0
torch=2.12.1+cu130 LayerNorm mode=autocast peak_gib=2.00 step_ms=7.67 kernels/step=19 cast_kernels/step=4
```

One norm costs +0.5 GiB peak (+34%): exactly one materialized fp32 copy of the (8, 4096, 4096)
bf16 activation (512 MiB), plus standalone cast kernels. The cost scales linearly with the number
of norm sites (see the at-scale numbers above). LayerNorm pays the same under autocast, confirming
the general Inductor behavior; for `rms_norm` this is new in 2.12 (no autocast registration on
2.11, where the two modes behave identically).

### Expected behavior
The `fp32` cast policy for `rms_norm` skips promotion when input and weight dtypes already
   match (the fused path dispatches fine and accumulates in fp32 internally)

### Workaround

Run the norm outside autocast; internal fp32 accumulation provides the same protection the policy
was adding externally:

```python
class AutocastExemptRMSNorm(nn.RMSNorm):
    def forward(self, x):
        with torch.autocast(x.device.type, enabled=False):
            return super().forward(x)
```

Measured at scale, this is faster and leaner than both the 2.12 default and the 2.11 baseline,
since the norm runs at bf16 I/O width on top of the improved 2.12 codegen.

### Versions

<details>
<summary>collect_env (regression env)</summary>

```
Collecting environment information...
PyTorch version: 2.12.1+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.39

Python version: 3.11.15 (main, Jun 23 2026, 15:19:43) [Clang 22.1.3 ] (64-bit runtime)
Python platform: Linux-6.12.68+-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 13.0.88
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA B200
GPU 1: NVIDIA B200
GPU 2: NVIDIA B200
GPU 3: NVIDIA B200
GPU 4: NVIDIA B200
GPU 5: NVIDIA B200
GPU 6: NVIDIA B200
GPU 7: NVIDIA B200

Nvidia driver version: 580.126.09
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.15.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  224
On-line CPU(s) list:                     0-223
Vendor ID:                               GenuineIntel
Model name:                              INTEL(R) XEON(R) PLATINUM 8581C CPU @ 2.10GHz
CPU family:                              6
Model:                                   207
Thread(s) per core:                      2
Core(s) per socket:                      56
Socket(s):                               2
Stepping:                                2
BogoMIPS:                                4200.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 wbnoinvd arat avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid cldemote movdiri movdir64b fsrm md_clear serialize tsxldtrk amx_bf16 avx512_fp16 amx_tile amx_int8 arch_capabilities
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               5.3 MiB (112 instances)
L1i cache:                               3.5 MiB (112 instances)
L2 cache:                                224 MiB (112 instances)
L3 cache:                                520 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-55,112-167
NUMA node1 CPU(s):                       56-111,168-223
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvidia-cublas==13.4.1.1
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] onnx==1.22.0
[pip3] onnx-ir==0.2.1
[pip3] onnxscript==0.7.0
[pip3] pytorch-lightning==2.6.5
[pip3] torch==2.12.1+cu130
[pip3] torch_c_dlpack_ext==0.1.5
[pip3] torchmetrics==1.9.0
[pip3] torchvision==0.27.1+cu130
[pip3] transformer_engine_torch==2.16.1
[pip3] triton==3.7.1
[conda] No relevant packages
```

</details>

Baseline env: torch 2.11.0+cu130, triton 3.6.0, CUDA 13.0, same machine.

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5 @chauhang @penguinwu @voznesenskym @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by abhishek-patnia-nu** [comment]
Does it make sense to move _(rms_norm) from AT_FORALL_FP32 (aten/src/ATen/autocast_mode.h L878 on main) to AT_FORALL_PROMOTE (L948). Matched dtypes should be safe to pass through, since the fused kernel already accumulates in fp32 internally and mixed dtypes are still promoted, so the #167308 fix should keep working. If yes, I can open a PR in the coming week.

**Comment by Sahil170595** [comment]
Confirmed the core regression locally on a different GPU/setup:

- Windows 11
- NVIDIA RTX 4080 Laptop GPU, SM 8.9
- torch 2.14.0.dev20260705+cu132
- triton-windows 3.7.1

Using the reported bf16 `Linear(4096,4096) -> RMSNorm(4096)` compiled forward+backward shape `(8, 4096, 4096)`:

- default autocast: `peak_gib=2.62`, `kernels/step=15`, `cast_kernels/step=3`
- RMSNorm under `autocast(enabled=False)`: `peak_gib=1.34`, `kernels/step=16`, `cast_kernels/step=0`

Also confirmed the eager autocast policy behavior directly: bf16 input + bf16 RMSNorm weight returns fp32 under autocast, but bf16 when autocast is disabled around the norm.
